ln-gateway/src/lib.rs

```diff
@@ -148,37 +146,77 @@ impl LnGateway {
+
+    async fn handle_htlc_incoming_msg(
+        &self,
+        htlc_accepted: HtlcAccepted,
+    ) -> Result<Preimage, LnGatewayError> {
+        let amount = htlc_accepted.htlc.amount;
+        let payment_hash = htlc_accepted.htlc.payment_hash;
+        let rng = rand::rngs::OsRng::new().unwrap();
+
+        tracing::debug!("Incoming htlc for payment hash {}", payment_hash);
+        let (outpoint, _) = self.buy_preimage_offer(&payment_hash, &amount, rng).await?;
+        tracing::debug!("Decrypting preimage {}", payment_hash);
+        let preimage = self.await_preimage_decryption(outpoint).await?;
+        tracing::debug!("Decrypted preimage {:?}", payment_hash);
+
+        Ok(preimage)
+    }
+
+    pub async fn run(&mut self) -> Result<(), LnGatewayError> {
+        // TODO: try to drive forward payments that were interrupted
+        loop {
+            let least_wait_until = Instant::now() + Duration::from_millis(100);
+            let pending_fetches = self
+                .federation_client
+                .list_fetchable_coins()
+                .into_iter()
+                .map(|out_point| {
+                    // TODO: get rid of cloning
+                    let federation_client = self.federation_client.clone();
+                    async move {
+                        if let Err(e) = federation_client.fetch_coins(out_point).await {
+                            debug!(error = %e, "Fetching coins failed");
+                        }
+                    }
+                })
+                .collect::<Vec<_>>();
+            futures::future::join_all(pending_fetches).await;
+
+            // Handle messages from webserver and plugin
+            while let Ok(msg) = self.receiver.try_recv() {
+                tracing::trace!("Gateway received message {:?}", msg);
+                match msg {
+                    GatewayRequest::HtlcAccpeted((htlc_accepted, sender)) => {
+                        let result = self.handle_htlc_incoming_msg(htlc_accepted).await;
+                        sender.send(result).expect("couldn't send over channel");
+                    }
+                    GatewayRequest::PayInvoice((contract_id, sender)) => {
+                        let result = self.handle_pay_invoice_msg(contract_id).await;
+                        sender.send(result).expect("couldn't send over channel");
                     }
                 }
-            })
-            .collect::<Vec<_>>();
-        futures::future::join_all(pending_fetches).await;
-        minimint_api::task::sleep_until(least_wait_until).await;
+            }
+
+            minimint_api::task::sleep_until(least_wait_until).await;
+        }
     }
 }
 
 impl Drop for LnGateway {
     fn drop(&mut self) {
-        self.fetcher.abort();
-        assert!(futures::executor::block_on(&mut self.fetcher).is_err());
+        self.webserver.abort();
+        // FIXME: this panics ...
+        // assert!(futures::executor::block_on(&mut self.webserver).is_err());
```

Comment:
This assertion fails cc @elsirion. Not sure how to correctly do this cleanup ...